File: frontend/apps/ios/Pantopus/Features/Scheduling/Business/Invoices/InvoicesListViewModel.swift
//
//  InvoicesListViewModel.swift
//  Pantopus
//
//  G12 Invoices List (owner, business-only) — Stream I15. Lists `GET /invoices`
//  grouped by created day, with the Stripe-not-connected gate from
//  `GET /payments/status`. Behind `SchedulingFeatureFlags.paidEnabled`. Matches
//  `invoiceslist-frames.jsx` (within the limits of the minimal InvoiceDTO — see
//  InvoicesKit note: no status pills / status filters / status summary).
//

import Foundation
import Observation
import SwiftUI

@Observable
@MainActor
final class InvoicesListViewModel {
    enum Phase: Equatable { case loading, loaded, empty, gate, error(String), comingSoon }

    /// Status filter chips (`invoiceslist-frames.jsx` `FilterChips`). The DTO
    /// carries no `status`, so selecting a status chip cannot yet filter the
    /// list (deferred); `.all` is the only chip with data behind it. The chip
    /// row + selection are rendered now so the structure matches the design.
    enum InvoiceFilter: String, CaseIterable, Identifiable {
        case all = "All"
        case paid = "Paid"
        case sent = "Sent"
        case overdue = "Overdue"
        case refunded = "Refunded"

        var id: String { rawValue }
    }

    // MARK: Inputs

    let owner: SchedulingOwner
    let push: @MainActor (SchedulingRoute) -> Void
    private let client: SchedulingClient

    // MARK: State

    private(set) var phase: Phase = .loading
    private(set) var invoices: [InvoiceDTO] = []
    private(set) var sections: [InvoiceDaySection] = []
    private(set) var paymentsConnected = false
    private(set) var paymentsApplicable = true

    /// Active status filter chip. View-only until the DTO carries `status`.
    var selectedFilter: InvoiceFilter = .all

    var theme: SchedulingIdentityTheme { owner.theme }
    var accent: Color { theme.accent }

    /// The two-stat summary structure (`invoiceslist-frames.jsx` `Summary`).
    /// The design's KPIs are Outstanding (amber when any invoice is overdue) /
    /// Collected · month — both need per-invoice `status` + `paid_at`, which the
    /// lean DTO omits. Until those land we surface the two DTO-derivable totals
    /// in the same two-column layout, and the overdue-amber treatment stays a
    /// view-only capability driven by `hasOverdue` (always false today).
    var totalLabel: String {
        SchedulingMoney.format(cents: invoices.reduce(0) { $0 + ($1.totalCents ?? 0) }, currency: invoices.first?.currency)
    }

    var countLabel: String { "\(invoices.count)" }

    /// Drives the summary's amber treatment (design `Summary overdue`). No
    /// `status` in the DTO yet → no invoice can be flagged overdue, so this is
    /// always false until the field lands (deferred).
    var hasOverdue: Bool { false }

    init(
        owner: SchedulingOwner,
        push: @escaping @MainActor (SchedulingRoute) -> Void,
        client: SchedulingClient
    ) {
        self.owner = owner
        self.push = push
        self.client = client
    }

    // MARK: Lifecycle

    func load() async {
        guard SchedulingFeatureFlags.paidEnabled else { phase = .comingSoon; return }
        phase = .loading
        do {
            if let status: PaymentsStatusDTO = try? await client.request(SchedulingEndpoints.paymentsStatus(owner: owner)) {
                paymentsConnected = status.connected
                paymentsApplicable = status.applicable
            }
            let result: InvoicesResponse = try await client.request(SchedulingEndpoints.getInvoices(owner: owner))
            invoices = result.invoices
            sections = InvoiceGrouping.byDay(invoices)
            if !invoices.isEmpty {
                phase = .loaded
            } else if paymentsApplicable, !paymentsConnected {
                phase = .gate
            } else {
                phase = .empty
            }
        } catch let error as SchedulingError {
            phase = .error(error.userMessage ?? "Couldn't load invoices.")
        } catch {
            phase = .error("Couldn't load invoices.")
        }
    }

    func refresh() async { await load() }

    // MARK: Actions

    func openInvoice(_ invoice: InvoiceDTO) { push(.invoiceDetail(owner: owner, invoiceId: invoice.id)) }
    func connectPayments() { push(.paymentsSetup(owner: owner)) }

    func selectFilter(_ filter: InvoiceFilter) { selectedFilter = filter }

    /// Top-bar search affordance (`invoiceslist-frames.jsx` `SearchBtn`). The
    /// design draws the glyph but wires no destination, and there is no invoice
    /// search route/endpoint yet — so this is a no-op placeholder (deferred)
    /// keeping the chrome structurally faithful.
    func search() {}

    // MARK: Row formatting

    func amount(_ invoice: InvoiceDTO) -> String {
        SchedulingMoney.format(cents: invoice.totalCents, currency: invoice.currency)
    }

    /// Short monospace reference from the invoice id (no invoice_number in DTO).
    func reference(_ invoice: InvoiceDTO) -> String {
        "INV-" + invoice.id.prefix(6).uppercased()
    }

    /// Service sub-label (`INV-… · {service}`) — derived from the first parsed
    /// line item. Falls back to "Service" when `line_items` carries no label.
    func service(_ invoice: InvoiceDTO) -> String {
        InvoiceParsing.lineItems(from: invoice.lineItems).first?.label ?? "Service"
    }

    /// Two-letter initials for the payer avatar disc. The DTO has no payer
    /// display name (only `recipient_user_id`), so we derive a stable two-char
    /// token from the invoice reference until the name lands (deferred).
    func payerInitials(_ invoice: InvoiceDTO) -> String {
        let token = (invoice.recipientUserId ?? invoice.id)
            .filter(\.isLetter)
            .prefix(2)
            .uppercased()
        return token.isEmpty ? "IN" : token
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/scheduling/invoices/InvoicesListViewModel.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.ui.screens.scheduling.invoices

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.scheduling.InvoiceDto
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.auth.AuthRepository
import app.pantopus.android.data.scheduling.SchedulingError
import app.pantopus.android.data.scheduling.SchedulingErrorDecoder
import app.pantopus.android.data.scheduling.SchedulingFeatureFlags
import app.pantopus.android.data.scheduling.SchedulingOwner
import app.pantopus.android.data.scheduling.SchedulingRepository
import app.pantopus.android.ui.screens.scheduling._shared.SchedulingPillar
import app.pantopus.android.ui.screens.scheduling._shared.SchedulingRoutes
import app.pantopus.android.ui.screens.scheduling._shared.pillar
import app.pantopus.android.ui.screens.scheduling.packages.PackagesMoney
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Status filter chips (`invoiceslist-frames.jsx` `FilterChips`). The DTO carries
 * no `status`, so selecting a status chip cannot yet filter the list (deferred);
 * `All` is the only chip with data behind it. The chip row + selection render
 * now so the structure matches the design.
 */
enum class InvoiceFilter(val label: String) {
    All("All"),
    Paid("Paid"),
    Sent("Sent"),
    Overdue("Overdue"),
    Refunded("Refunded"),
}

/** G12 Invoices List UI state. */
sealed interface InvoicesListUiState {
    data object Loading : InvoicesListUiState

    data object ComingSoon : InvoicesListUiState

    /** Stripe not connected → "Connect payments to invoice" gate. */
    data object Gate : InvoicesListUiState

    data object Empty : InvoicesListUiState

    data class Error(val message: String) : InvoicesListUiState

    data class Loaded(
        val sections: List<InvoiceDaySection>,
        /** Formatted outstanding (unpaid) amount — "Outstanding" KPI left column. */
        val outstandingLabel: String,
        /** Formatted collected-this-month amount — "Collected · month" KPI right column. */
        val collectedMonthLabel: String,
        /**
         * True when at least one invoice is overdue — drives amber tint on the
         * "Outstanding" label + value. Always false until the DTO gains a `status`
         * field (deferred: InvoiceDto carries no `status` / `paid_at`).
         */
        val hasOverdue: Boolean,
        val pillar: SchedulingPillar,
    ) : InvoicesListUiState
}

/**
 * G12 Invoices List (owner, business-only) — Stream A15. Lists `GET /invoices`
 * grouped by created day, with the Stripe-not-connected gate from
 * `GET /payments/status`. Behind [SchedulingFeatureFlags]. Mirrors iOS
 * `InvoicesListViewModel` / `invoiceslist-frames.jsx` (within the minimal
 * InvoiceDto — no status pills / status filters with data / overdue summary).
 */
@HiltViewModel
class InvoicesListViewModel
    @Inject
    constructor(
        private val repo: SchedulingRepository,
        private val auth: AuthRepository,
        private val errors: SchedulingErrorDecoder,
        private val flags: SchedulingFeatureFlags,
    ) : ViewModel() {
        private val _state = MutableStateFlow<InvoicesListUiState>(InvoicesListUiState.Loading)
        val state: StateFlow<InvoicesListUiState> = _state.asStateFlow()

        private val _filter = MutableStateFlow(InvoiceFilter.All)
        val filter: StateFlow<InvoiceFilter> = _filter.asStateFlow()

        private var owner: SchedulingOwner = SchedulingOwner.Personal
        private var started = false

        fun start() {
            if (started) {
                refresh()
            } else {
                started = true
                owner = resolveOwner()
                load()
            }
        }

        fun load() {
            viewModelScope.launch {
                if (!flags.paidSchedulingEnabled) {
                    _state.value = InvoicesListUiState.ComingSoon
                    return@launch
                }
                _state.value = InvoicesListUiState.Loading
                val status = (repo.getPaymentsStatus(owner) as? NetworkResult.Success)?.data
                val connected = status?.connected ?: false
                val applicable = status?.applicable ?: true
                when (val result = repo.getInvoices(owner)) {
                    is NetworkResult.Success -> {
                        val invoices = result.data.invoices
                        _state.value =
                            when {
                                invoices.isNotEmpty() -> {
                                    // The DTO has no `status` or `paid_at` field, so we
                                    // cannot compute the real outstanding vs collected split.
                                    // Outstanding = sum of all invoice totals (deferred: should
                                    // exclude paid invoices once DTO gains `status`).
                                    // Collected · month = $0 placeholder (deferred: needs
                                    // `paid_at` to filter to current calendar month).
                                    val currency = invoices.firstOrNull()?.currency
                                    InvoicesListUiState.Loaded(
                                        sections = InvoiceGrouping.byDay(invoices),
                                        outstandingLabel =
                                            PackagesMoney.format(
                                                invoices.sumOf { it.totalCents ?: 0 },
                                                currency,
                                            ),
                                        collectedMonthLabel =
                                            PackagesMoney.format(0, currency),
                                        // Deferred: always false until DTO has `status`
                                        hasOverdue = false,
                                        pillar = owner.pillar(),
                                    )
                                }
                                applicable && !connected -> InvoicesListUiState.Gate
                                else -> InvoicesListUiState.Empty
                            }
                    }
                    is NetworkResult.Failure ->
                        _state.value =
                            InvoicesListUiState.Error(
                                errors.decode(result.error).message(),
                            )
                }
            }
        }

        fun refresh() = load()

        fun selectFilter(target: InvoiceFilter) {
            _filter.value = target
        }

        fun invoiceRoute(invoiceId: String): String = SchedulingRoutes.invoiceDetail(invoiceId)

        fun connectRoute(): String = SchedulingRoutes.PAYMENTS_SETUP

        // ─── Row formatting ─────────────────────────────────────────────────────

        fun amount(invoice: InvoiceDto): String = PackagesMoney.format(invoice.totalCents, invoice.currency)

        /** Short monospace reference from the invoice id (no invoice_number in DTO). */
        fun reference(invoice: InvoiceDto): String = "INV-" + invoice.id.take(REF_PREFIX_LEN).uppercase()

        /** Service sub-label — first parsed line item, else "Service". */
        fun service(invoice: InvoiceDto): String = InvoiceParsing.lineItems(invoice.lineItems).firstOrNull()?.label ?: "Service"

        /** Two-letter payer initials (no payer display name in the DTO). */
        fun payerInitials(invoice: InvoiceDto): String {
            val token = (invoice.recipientUserId ?: invoice.id).filter { it.isLetter() }.take(INITIALS_LEN).uppercase()
            return token.ifEmpty { "IN" }
        }

        private fun resolveOwner(): SchedulingOwner =
            (auth.state.value as? AuthRepository.State.SignedIn)?.user?.id
                ?.let { SchedulingOwner.Business(it) }
                ?: SchedulingOwner.Personal

        private fun SchedulingError.message(): String =
            when (this) {
                is SchedulingError.Secret -> "Only the business owner can view invoices."
                is SchedulingError.Generic -> message
                else -> "Couldn't load invoices."
            }

        private companion object {
            const val REF_PREFIX_LEN = 6
            const val INITIALS_LEN = 2
        }
    }
